File: 13RomanToInteger/RomanToInteger.java
package test;

import java.util.HashMap;
import java.util.Map;

/**
 * Created by xiao on 9/26/16.
 */
public class RomanToInteger {
    public static void main(String[] args){
        String ro = "MCMXCVI";
        System.out.println(ro + "to integer is " + romanToInt(ro));
    }

    public static int romanToInt(String s){
        int len = s.length();
        if (s == null || len == 0) {
            return 0;
        }
        Map<Character, Integer> map = new HashMap<Character, Integer>(){{
            put('I', 1);
            put('V', 5);
            put('X', 10);
            put('L', 50);
            put('C', 100);
            put('D', 500);
            put('M', 1000);

        }};

        int result = map.get(s.charAt(len - 1));
        for (int i = len - 2 ; i >= 0; i--)
            if (map.get(s.charAt(i + 1)).intValue() > map.get(s.charAt(i)).intValue()) {
                result -= map.get(s.charAt(i));
            } else {
                result += map.get(s.charAt(i));
            }
        return result;
    }
}

File: 13RomanToInteger/RomanToInteger.c
/*************************************************************************
    > File Name: RomanToInteger.c
    > Author: zhang zheng
    > E-mail: kezy-5566@163.com 
    > Created Time: Mon 26 Sep 2016 05:19:47 PM CST
    Roman to Integer:
    	Given a roman numeral, convert it to an integer.
    	Input is guaranteed to be within the range from 1 to 3999.
 ************************************************************************/

#include <stdio.h>
#include <string.h>

int romanToInt(char *s)
{
	int len = strlen(s);
	if (strlen == 0)
		return 0;
	int map[26] = {0};
	map['I' - 'A'] = 1;
	map['V' - 'A'] = 5;
	map['X' - 'A'] = 10;
	map['L' - 'A'] = 50;
	map['C' - 'A'] = 100;
	map['D' - 'A'] = 500;
	map['M' - 'A'] = 1000;

	int result = map[*(s+len-1) - 'A'];
	for (int i = len - 2; i >= 0; i--) {
		if (map[*(s + i + 1) - 'A'] > map[*(s + i) - 'A']){
			result -= map[*(s + i) - 'A'];
		} else {
			result += map[*(s + i) - 'A'];
		}
	}
	printf("%d\n", result);
	return result;
	
}

int main()
{
	char *s = "MCMXCVI";
	int num = romanToInt(s);
	return 0;
}


